File: docs/fontlib.js
// Port of fontlib.py: enough sfnt parsing to name a font and measure what it
// really covers. Runs entirely in the browser, nothing is uploaded.

const UPPER = range(65, 90), LOWER = range(97, 122), DIGITS = range(48, 57);
const PUNCT = [...".,!?'\"-:;()"].map(c => c.charCodeAt(0));

function range(a, b) { const o = []; for (let i = a; i <= b; i++) o.push(i); return o; }

function tables(dv, offset = 0) {
  const tag = dv.getUint32(offset);
  if (tag === 0x74746366) return tables(dv, dv.getUint32(offset + 12)); // 'ttcf'
  if (![0x4f54544f, 0x00010000, 0x74727565, 0x74797031].includes(tag))
    throw new Error('not a font file');
  const count = dv.getUint16(offset + 4), out = {};
  for (let i = 0; i < count; i++) {
    const rec = offset + 12 + 16 * i;
    if (rec + 16 > dv.byteLength) break;
    let name = '';
    for (let j = 0; j < 4; j++) name += String.fromCharCode(dv.getUint8(rec + j));
    out[name] = [dv.getUint32(rec + 8), dv.getUint32(rec + 12)];
  }
  return out;
}

function names(dv, tabs) {
  if (!tabs.name) return {};
  const off = tabs.name[0];
  const count = dv.getUint16(off + 2), storage = off + dv.getUint16(off + 4);
  const best = {};
  for (let i = 0; i < count; i++) {
    const rec = off + 6 + 12 * i;
    if (rec + 12 > dv.byteLength) break;
    const pid = dv.getUint16(rec), nid = dv.getUint16(rec + 6);
    const len = dv.getUint16(rec + 8), noff = dv.getUint16(rec + 10);
    const bytes = new Uint8Array(dv.buffer, dv.byteOffset + storage + noff, len);
    let value = '';
    if (pid === 3) {
      for (let j = 0; j + 1 < len; j += 2) value += String.fromCharCode((bytes[j] << 8) | bytes[j + 1]);
    } else {
      for (let j = 0; j < len; j++) value += String.fromCharCode(bytes[j]);
    }
    value = value.replace(/\0/g, '').trim();
    if (!value) continue;
    if (!(nid in best) || pid === 3) best[nid] = value;
  }
  return best;
}

function coverage(dv, tabs) {
  if (!tabs.cmap) return new Set();
  const off = tabs.cmap[0], n = dv.getUint16(off + 2);
  let chosen = null, fallback = null;
  for (let i = 0; i < n; i++) {
    const rec = off + 4 + 8 * i;
    const pid = dv.getUint16(rec), eid = dv.getUint16(rec + 2), sub = dv.getUint32(rec + 4);
    if ((pid === 3 && eid === 10) || (pid === 0 && (eid === 4 || eid === 6))) { chosen = off + sub; break; }
    if ((pid === 3 && eid === 1) || (pid === 0 && eid === 3)) chosen = chosen ?? off + sub;
    else if (fallback === null) fallback = off + sub;
  }
  const start = chosen ?? fallback;
  if (start === null) return new Set();
  const fmt = dv.getUint16(start), chars = new Set();
  if (fmt === 4) {
    const segX2 = dv.getUint16(start + 6), segs = segX2 / 2;
    const endO = start + 14, startO = endO + segX2 + 2;
    for (let i = 0; i < segs; i++) {
      const end = dv.getUint16(endO + 2 * i), beg = dv.getUint16(startO + 2 * i);
      if (beg === 0xffff) continue;
      for (let c = beg; c <= Math.min(end, 0xfffe); c++) chars.add(c);
    }
  } else if (fmt === 12) {
    const groups = dv.getUint32(start + 12);
    for (let i = 0; i < groups; i++) {
      const rec = start + 16 + 12 * i;
      if (rec + 12 > dv.byteLength) break;
      const beg = dv.getUint32(rec), end = dv.getUint32(rec + 4);
      if (end - beg > 0x10ffff) continue;
      for (let c = beg; c <= Math.min(end, 0x10ffff); c++) chars.add(c);
    }
  } else if (fmt === 6) {
    const first = dv.getUint16(start + 6), cnt = dv.getUint16(start + 8);
    for (let c = first; c < first + cnt; c++) chars.add(c);
  } else if (fmt === 0) {
    for (let c = 0; c < 256; c++) if (dv.getUint8(start + 6 + c)) chars.add(c);
  }
  return chars;
}

export function describe(buffer, filename = '') {
  const dv = new DataView(buffer);
  const tabs = tables(dv);
  const nm = names(dv, tabs), chars = coverage(dv, tabs);
  const truetype = 'glyf' in tabs;
  const ext = filename.includes('.') ? filename.split('.').pop().toLowerCase() : '';
  const flags = [];
  if ((ext === 'ttf' && !truetype) || (ext === 'otf' && truetype)) flags.push('mislabelled');
  const missing = [];
  const has = set => set.every(c => chars.has(c));
  if (!has(UPPER)) missing.push('A-Z');
  if (!has(LOWER)) missing.push('a-z');
  if (!has(DIGITS)) missing.push('0-9');
  if (!has(PUNCT)) missing.push('punctuation');
  if (missing.length) flags.push('incomplete');
  const family = nm[16] || nm[1] || filename;
  const ps = nm[6] || '';
  const blob = `${ps} ${nm[4] || ''} ${family} ${filename}`.toLowerCase();
  if (/demo|personal use|trial/.test(blob)) flags.push('demo');
  return {
    filename, family, style: nm[17] || nm[2] || 'Regular', psname: ps,
    full: nm[4] || `${family} ${nm[2] || ''}`.trim(),
    outlines: truetype ? 'TrueType' : 'CFF',
    chars: chars.size, missing, flags, size: buffer.byteLength,
  };
}

File: fontlib.py
"""Font inspection helpers. Stdlib only.

Parses sfnt containers (OTF/TTF/TTC) far enough to answer the two questions
that actually matter when deciding what to install: what is this font called,
and how much of the alphabet does it really cover.
"""
import struct

MAC_ROMAN = "mac-roman"


class NotAFont(Exception):
    pass


def _tables(buf, offset=0):
    if len(buf) < offset + 12:
        raise NotAFont("too short")
    tag = buf[offset:offset + 4]
    if tag == b"ttcf":
        first = struct.unpack(">I", buf[offset + 12:offset + 16])[0]
        return _tables(buf, first)
    if tag not in (b"OTTO", b"\x00\x01\x00\x00", b"true", b"typ1"):
        raise NotAFont("unknown sfnt tag %r" % tag)
    count = struct.unpack(">H", buf[offset + 4:offset + 6])[0]
    out = {}
    for i in range(count):
        rec = offset + 12 + 16 * i
        if rec + 16 > len(buf):
            break
        name = buf[rec:rec + 4].decode("latin-1")
        start, length = struct.unpack(">II", buf[rec + 8:rec + 16])
        out[name] = (start, length)
    return out


def sfnt_tag(buf):
    return buf[:4]


def is_truetype_outlines(buf):
    """True when the file carries glyf outlines rather than CFF."""
    try:
        return "glyf" in _tables(buf)
    except NotAFont:
        return False


def names(buf):
    """Return {nameID: string} preferring the Windows/Unicode records."""
    tabs = _tables(buf)
    if "name" not in tabs:
        return {}
    off, _ = tabs["name"]
    if off + 6 > len(buf):
        return {}
    count, str_off = struct.unpack(">HH", buf[off + 2:off + 6])
    storage = off + str_off
    best = {}
    for i in range(count):
        rec = off + 6 + 12 * i
        if rec + 12 > len(buf):
            break
        pid, eid, lid, nid, length, noff = struct.unpack(">HHHHHH", buf[rec:rec + 12])
        raw = buf[storage + noff:storage + noff + length]
        if not raw:
            continue
        try:
            value = raw.decode("utf-16-be") if pid == 3 else raw.decode(MAC_ROMAN)
        except (UnicodeDecodeError, LookupError):
            continue
        value = value.strip("\x00").strip()
        if not value:
            continue
        # Windows records win; a Mac record only fills a gap.
        if nid not in best or pid == 3:
            best[nid] = value
    return best


def coverage(buf):
    """Set of Unicode codepoints the font claims to support."""
    tabs = _tables(buf)
    if "cmap" not in tabs:
        return set()
    off, _ = tabs["cmap"]
    if off + 4 > len(buf):
        return set()
    n = struct.unpack(">H", buf[off + 2:off + 4])[0]
    chosen = None
    fallback = None
    for i in range(n):
        rec = off + 4 + 8 * i
        if rec + 8 > len(buf):
            break
        pid, eid, sub = struct.unpack(">HHI", buf[rec:rec + 8])
        if (pid, eid) in ((3, 10), (0, 4), (0, 6)):
            chosen = off + sub
            break
        if (pid, eid) in ((3, 1), (0, 3)):
            chosen = chosen or off + sub
        elif fallback is None:
            fallback = off + sub
    start = chosen if chosen is not None else fallback
    if start is None or start + 2 > len(buf):
        return set()
    fmt = struct.unpack(">H", buf[start:start + 2])[0]
    chars = set()
    if fmt == 4:
        seg_x2 = struct.unpack(">H", buf[start + 6:start + 8])[0]
        segs = seg_x2 // 2
        end_o = start + 14
        start_o = end_o + seg_x2 + 2
        for i in range(segs):
            end = struct.unpack(">H", buf[end_o + 2 * i:end_o + 2 * i + 2])[0]
            beg = struct.unpack(">H", buf[start_o + 2 * i:start_o + 2 * i + 2])[0]
            if beg == 0xFFFF:
                continue
            chars.update(range(beg, min(end, 0xFFFE) + 1))
    elif fmt == 12:
        groups = struct.unpack(">I", buf[start + 12:start + 16])[0]
        for i in range(groups):
            rec = start + 16 + 12 * i
            if rec + 12 > len(buf):
                break
            beg, end, _gid = struct.unpack(">III", buf[rec:rec + 12])
            if end - beg > 0x10FFFF:
                continue
            chars.update(range(beg, min(end, 0x10FFFF) + 1))
    elif fmt == 6:
        first, cnt = struct.unpack(">HH", buf[start + 6:start + 10])
        chars.update(range(first, first + cnt))
    elif fmt == 0:
        for code in range(256):
            if buf[start + 6 + code]:
                chars.add(code)
    return chars


UPPER = set(range(ord("A"), ord("Z") + 1))
LOWER = set(range(ord("a"), ord("z") + 1))
DIGITS = set(range(ord("0"), ord("9") + 1))
PUNCT = {ord(c) for c in ".,!?'\"-:;()"}


def describe(buf, filename=""):
    """Everything the dashboard needs to judge one font file."""
    nm = names(buf)
    chars = coverage(buf)
    family = nm.get(16) or nm.get(1) or ""
    style = nm.get(17) or nm.get(2) or ""
    ps = nm.get(6) or ""
    full = nm.get(4) or (family + " " + style).strip()
    tag = sfnt_tag(buf)
    truetype = is_truetype_outlines(buf)
    ext = filename.rsplit(".", 1)[-1].lower() if "." in filename else ""
    flags = []
    if ext == "ttf" and not truetype:
        flags.append("mislabelled")
    if ext == "otf" and truetype:
        flags.append("mislabelled")
    missing = []
    if not UPPER <= chars:
        missing.append("A-Z")
    if not LOWER <= chars:
        missing.append("a-z")
    if not DIGITS <= chars:
        missing.append("0-9")
    if not PUNCT <= chars:
        missing.append("punctuation")
    if missing:
        flags.append("incomplete")
    blob = (ps + " " + full + " " + family + " " + filename).lower()
    if "demo" in blob or "personal use" in blob or "trial" in blob:
        flags.append("demo")
    return {
        "family": family or (ps or filename),
        "style": style,
        "psname": ps,
        "full": full,
        "outlines": "TrueType" if truetype else "CFF",
        "sfnt": tag.decode("latin-1", "replace"),
        "chars": len(chars),
        "missing": missing,
        "flags": flags,
        "size": len(buf),
    }
